### PR TITLE
Add a f16 option to the NN example

### PR DESCRIPTION
**Pull request by WilliamTambellini**
Add a f16 option to the NN example:
examples/machine_learning/neural_network_cuda 0 - 60 f16
f32:
Training set: Accuracy on training data: 98.64
Test set: Accuracy on testing  data: 93.29
f16:
Training set: Accuracy on training data: 98.94
Test set: Accuracy on testing  data: 93.50

### CODE REVIEW

**Comment by WilliamTambellini** [comment]
@umar456 @9prady9   : that one is an easy way to play with f16 and a clear evidence that AF can now run some f16 ops. Could that one be considered/merged before 3.7.0 release ?

File: examples/machine_learning/neural_network.cpp [hunk 1]
```diff
@@ -18,6 +18,16 @@
 using namespace af;
 using std::vector;
 
+#define ASSERTLOG(TEST, MESSAGE) if (!(TEST)) std::cerr << MESSAGE << std::endl;
```
Comment (umar456) [inline examples/machine_learning/neural_network.cpp]: Could you remove the ASSERTLOG from this example. I don't think we should have these messages in the examples.

File: examples/machine_learning/neural_network.cpp [hunk 2]
```diff
@@ -84,30 +97,39 @@ void ann::back_propagate(const vector<array> signal, const array &target,
...
     }
 }
 
-ann::ann(vector<int> layers, double range)
-    : num_layers(layers.size()), weights(layers.size() - 1) {
-    // Generate uniformly distributed random numbers between [-range/2,range/2]
+
+ann::ann(vector<int> layers, double range, dtype dt)
+    : num_layers(layers.size()), weights(layers.size() - 1), datatype(dt) {
+    std::cout << "Initializing weights using a random uniformly distribution between " << -range/2 << " and " << range/2 << " at precision " << toStr(datatype) << std::endl;
```
Comment (umar456) [inline examples/machine_learning/neural_network.cpp]: This should fit between 80 columns.

File: examples/machine_learning/neural_network.cpp (hunk 1 above)
Comment (WilliamTambellini) [reply examples/machine_learning/neural_network.cpp]: done

**Comment by WilliamTambellini** [comment]
hum
12/126 Test  #13: test_canny_cpu ...................***Failed    0.34 sec
on Macos.
